WalletWasabi/Crypto/ZeroKnowledge/Challenge.cs

```diff
@@ -0,0 +1,42 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class Challenge
+	{
+		public static Scalar Build(GroupElement publicPoint, GroupElement nonce, IEnumerable<GroupElement> generators)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			Guard.False($"{nameof(nonce)}.{nameof(nonce.IsInfinity)}", nonce.IsInfinity);
+			foreach (var generator in generators)
+			{
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+			}
+
+			if (publicPoint == nonce)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(nonce)} should not be equal.");
+			}
+
+			return HashToScalar(new[] { publicPoint, nonce }.Concat(generators));
+		}
+
+		public static Scalar Build(GroupElement publicPoint, GroupElement nonce, params GroupElement[] generators)
+			=> Build(publicPoint, nonce, generators as IEnumerable<GroupElement>);
+
+		private static Scalar HashToScalar(IEnumerable<GroupElement> transcript)
+		{
+			var concatenation = ByteHelpers.Combine(transcript.Select(x => x.ToBytes()));
```

Comment:
unless everything can always guaranteed to be fixed size (including any future things), concatenation should probably use some sort of tagging & length encoding, see the Merlin transcript docs for one approach (using STROBE protocol framework. a fair tl;dr is that it uses hash trees to ensure boundaries between elements are well defined, not just their order)

Reply:
I'm sorry, but I don't udnerstand.

Reply:
It's important to ensure the transcripts are committed fully and unambiguously for the strong variant of Fiat-Shamir vairant. This categorically eliminates a class of attacks, and is easy to achieve. If strings are just concatenated then in principle such an attack would be possible (not with these arguments since they are all fixed length and in a set order, but in general).

Merlin is a generic ZK proof library in Rust that uses STROBE for the transcript hashes. [STROBE](https://strobe.sourceforge.io/) uses SHAKE, a variant of SHA-3 to do this kind of thing generically with very little code, so it's a safe and standard way of encoding heterogenous, potentially mixed length data streams into a transcript and ensuring both sides are synced up with the same hash.

By borrowing this approach (again see previous comment, it doesn't matter in this PR) we can be more confident that none of our proofs are accidentally weak FS, and that the whole framework could be extended possibly by others without creating compatibility issues, as well as safely compose compound proofs out of basic ones (using such an approach gives conjunctions for free basically)

Reply:
short explanation of the kind of issues with unstructured concatenation at 27:42: https://youtu.be/BJnjAF2cz48?t=1662

Reply:
Very interesting. We don't need to hash this with their length because we work with fixed-length FE (they are serialized as 33 bytes length always) and I cannot imagine any way an attacker can use that against us. **However** it is highly probable we need to hash some other info in as part of WabiSabi implementation (or future extensions) and I think it is important to eliminate the possibility of these kind of attacks.

I am going to create a PR for this, i think the additional complexity is negligible in comparison with the security improvement.

Reply:
Done https://github.com/nopara73/WalletWasabi/pull/33

Reply:
https://github.com/nopara73/WalletWasabi/pull/33#issuecomment-674786255